content/blog/nats-operator-and-service-accounts.md

```diff
@@ -0,0 +1,337 @@
+
+```yaml
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nats-user-pod
+  labels:
+    nats_cluster: example-nats
+spec:
+  volumes:
+    - name: "token"
+      projected:
+        sources:
+        - secret:
+            name: "nats-user-example-nats-bound-token"
+            items:
+              - key: token
+                path: "token"
+  restartPolicy: Never
+  containers:
+    - name: nats-ops
+      command: ["/bin/sh"]
+      image: "wallyqs/nats-ops:latest"
+      tty: true
+      stdin: true
+      stdinOnce: true
+
+      # Service Account Token is mounted via projected volume.
+      volumeMounts:
+      - name: "token"
+        mountPath: "/var/run/secrets/nats.io"
+        readOnly: true
+```
+
+Using `kubectl attach` we can run a few commands in that container to confirm
+that the permissions have been applied to the cluster:
+
+```sh
+kubectl -n nats-io attach -it nats-user-pod
+...
+/go # nats-sub -s nats://nats-user:`cat /var/run/secrets/nats.io/token`@example-nats:4222 hello.world
+Listening on [hello.world]
+^C
+^C
+/go # nats-sub -s nats://nats-user:`cat /var/run/secrets/nats.io/token`@example-nats:4222 hi
+nats: permissions violation for subscription to "hi"
+/go # nats-sub -s nats://nats-user:`cat /var/run/secrets/nats.io/token`@example-nats:4222 foo.bar
+Listening on [foo.bar]
+```
+
+Since config reload was enabled, at any time we can update the
+authorization configuration for the cluster and modify the
+permissions.  For example, if the `NatsServiceRole` is deleted while a
+client is connected, then the client will be disconnected from the
+cluster since the account is no longer valid:
+
+```sh
+$ kubectl -n nats-io delete natsservicerole nats-user
+natsservicerole.nats.io "nats-user" deleted
+
+# (in the attached container)
+/go # nats-sub -s nats://nats-user:`cat /var/run/secrets/nats.io/token`@example-nats:4222 hello.world
+Listening on [hello.world]
+Got disconnected!
+```
+
+Similarly, we can modify the permissions in the `NatsServiceRole` and the servers will update
+the permissions and reload. Config reload events can be confirmed from the logs from a NATS server.
+
+```sh
+kubectl -n nats-io logs example-nats-1 nats
+...
+[6] 2018/07/19 01:05:43.469246 [INF] Reloaded server configuration
+```
+
+### How it works
+
+As previously mentioned, the current implementation of the integration
+relies on the `TokenRequest` and `PodShareProcessNamespace` alpha features
+that were made available in the Kubernetes v1.10 release.
+
+The [TokenRequest API](https://github.com/kubernetes/community/blob/4a41674f6fb6fa69b3dfaf15267d908e909685d7/contributors/design-proposals/auth/bound-service-account-tokens.md) feature which allows issuing new tokens related
+to a `ServiceAccount` but with a different audience as the one used for
+the API server.  _Customizing the audience_ is very important in this
+case, since using the bare original token from `ServiceAccount` can
```

Comment:
The [TokenRequest API](https://github.com/kubernetes/community/blob/4a41674f6fb6fa69b3dfaf15267d908e909685d7/contributors/design-proposals/auth/bound-service-account-tokens.md) feature allows the issuance of new tokens related to a `ServiceAccount` but with a different audience as the one used for the API server.